README.md
# jmespath.js

A JavaScript implementation of JMESPath, which is a query language for JSON. It will take a JSON
document and transform it into another JSON document through a JMESPath expression. This fork was
originally based from the [daz-is/jmespath.js](https://github.com/daz-is/jmespath.js) fork, which
is highly recommended to leverage instead of this project. This fork exists for strict compliance,
security, and organizational feature deviation purposes alone.

```js
const jmespath = require('jmespath');

jmespath.search({foo: {bar: {baz: [1, 2, 3]}}}, 'foo.bar.baz[2]')
```

> `3`


## Installation

```console
$ npm install --save @cloudelements/jmespath
```

## Adding custom functions

Custom functions can be added to the JMESPath runtime by using the `decorate` function:

```js
function customFunc(resolvedArgs) {
  return resolvedArgs[0] + 99;
}

const extraFunctions = {
  custom: {_func: customFunc, _signature: [{types: [jmespath.types.TYPE_NUMBER]}]},
};

jmespath.decorate(extraFunctions);
```

The value returned by the decorate function is a curried function (takes arguments one at a time)
that takes the search expression first and then the data to search against as the second parameter:

```js
jmespath.decorate(extraFunctions)('custom(`1`)')({})
```

> `100`

Because the return value from `decorate` is a curried function the result of compiling the
expression can be cached and run multiple times against different data:

```js
const expr = jmespath.decorate({})('a');
let value;

value = expr({a: 1});
assert.strictEqual(value, 1);

value = expr({a: 2});
assert.strictEqual(value, 2);
```


CHANGELOG.md
# [0.23.0](https://github.com/cloud-elements/jmespath.js/compare/v0.22.2...v0.23.0) (2023-08-03)


### Features

* add a remove function ([2f7e57e](https://github.com/cloud-elements/jmespath.js/commit/2f7e57e881fbfefb5699ecd032911b49ffaad60e))



## [0.22.2](https://github.com/cloud-elements/jmespath.js/compare/v0.22.1...v0.22.2) (2022-12-15)


### Bug Fixes

* non-scalar types should be made case insensitive with case insensitive flag (ENGCE-23763) ([#16](https://github.com/cloud-elements/jmespath.js/issues/16)) ([f659573](https://github.com/cloud-elements/jmespath.js/commit/f6595733a8e82f4f67416b23ece2a3adfb838e5f))



## [0.22.1](https://github.com/cloud-elements/jmespath.js/compare/v0.22.0...v0.22.1) (2022-11-30)


### Bug Fixes

* update TS signature for search to include opts ([#15](https://github.com/cloud-elements/jmespath.js/issues/15)) ([7e49b3b](https://github.com/cloud-elements/jmespath.js/commit/7e49b3bb3814a1faeb96519d0834ed1a35b52984))



# [0.22.0](https://github.com/cloud-elements/jmespath.js/compare/v0.21.0...v0.22.0) (2022-11-30)


### Features

* support case insensitive comparisons (ENGCE-22058) ([#14](https://github.com/cloud-elements/jmespath.js/issues/14)) ([a66be83](https://github.com/cloud-elements/jmespath.js/commit/a66be83f63601e170e68fe17cf8fc52524a7df0c))



# [0.21.0](https://github.com/cloud-elements/jmespath.js/compare/v0.20.0...v0.21.0) (2022-10-18)


### Bug Fixes

* address starts_with and ends_with edge cases (ENGCE-21162) ([#12](https://github.com/cloud-elements/jmespath.js/issues/12)) ([d4bf3b9](https://github.com/cloud-elements/jmespath.js/commit/d4bf3b9381c1fb033e50fddd697d8680ea8ea30d))


### Features

* add mult and div custom functions ([#11](https://github.com/cloud-elements/jmespath.js/issues/11)) ([4e23f14](https://github.com/cloud-elements/jmespath.js/commit/4e23f1480575b4d7a2f8668bed6e3b3e0d58c956))



# [0.20.0](https://github.com/cloud-elements/jmespath.js/compare/v0.19.0...v0.20.0) (2022-07-01)


### Features

* add pola-mode for comparators ([#9](https://github.com/cloud-elements/jmespath.js/issues/9)) ([ebd2655](https://github.com/cloud-elements/jmespath.js/commit/ebd2655650c97b7a11f3caf998ad7e9c050db4e0))
* pola add, contains, ends_with, length, starts_with, sub fns ([#10](https://github.com/cloud-elements/jmespath.js/issues/10)) ([8918107](https://github.com/cloud-elements/jmespath.js/commit/8918107d28a69807d3bc79d186cac6a3f4053fd1))



# [0.19.0](https://github.com/cloud-elements/jmespath.js/compare/v0.18.0...v0.19.0) (2021-11-23)


### Features

* add support in to_number for date strings (ENG-9197) ([#5](https://github.com/cloud-elements/jmespath.js/issues/5)) ([7b07df5](https://github.com/cloud-elements/jmespath.js/commit/7b07df5103e64c4e8d4c1dc62f293dacd8c32c23))
* support custom function add (ENG-9197) ([#6](https://github.com/cloud-elements/jmespath.js/issues/6)) ([36b9ca5](https://github.com/cloud-elements/jmespath.js/commit/36b9ca550510896062ff42d1294c12fbb56c4d83))
* support custom function sub (ENG-9197) ([#7](https://github.com/cloud-elements/jmespath.js/issues/7)) ([fb62219](https://github.com/cloud-elements/jmespath.js/commit/fb62219aaab50f4960797c0fc451bdd6be62d53e))



# [0.17.0](https://github.com/cloud-elements/jmespath.js/compare/v0.16.0...v0.17.0) (2020-03-09)


### Features

* add minimal TypeScript definitions ([#3](https://github.com/cloud-elements/jmespath.js/issues/3)) (CU-582) ([51b422d](https://github.com/cloud-elements/jmespath.js/commit/51b422dffafc16cc98e00e773dc78a4a29811cb9))



# [0.16.0](https://github.com/cloud-elements/jmespath.js/compare/0.15.0...0.16.0) (2020-03-09)

### Features

* clean up and pull under cloudelements ([#1](https://github.com/cloud-elements/jmespath.js/issues/1)) (CU-582) ([0acb1f4](https://github.com/cloud-elements/jmespath.js/commit/0acb1f45f0f27da3051d459cbebb73a9038a56db))


test/jmespath.js
var assert = require('assert');
var jmespath = require('../src/jmespath');
var tokenize = jmespath.tokenize;
var compile = jmespath.compile;
var strictDeepEqual = jmespath.strictDeepEqual;


describe('tokenize', function() {
    it('should tokenize unquoted identifier', function() {
        assert.deepEqual(tokenize('foo'),
                         [{type: "UnquotedIdentifier",
                          value: "foo",
                          start: 0}]);
    });
    it('should tokenize unquoted identifier with underscore', function() {
        assert.deepEqual(tokenize('_underscore'),
                          [{type: "UnquotedIdentifier",
                           value: "_underscore",
                           start: 0}]);
    });
    it('should tokenize unquoted identifier with numbers', function() {
        assert.deepEqual(tokenize('foo123'),
                          [{type: "UnquotedIdentifier",
                           value: "foo123",
                           start: 0}]);
    });
    it('should tokenize dotted lookups', function() {
        assert.deepEqual(
            tokenize('foo.bar'),
            [{type: "UnquotedIdentifier", value: "foo", start: 0},
             {type: "Dot", value: ".", start: 3},
             {type: "UnquotedIdentifier", value: "bar", start: 4},
            ]);
    });
    it('should tokenize numbers', function() {
        assert.deepEqual(
            tokenize('foo[0]'),
            [{type: "UnquotedIdentifier", value: "foo", start: 0},
             {type: "Lbracket", value: "[", start: 3},
             {type: "Number", value: 0, start: 4},
             {type: "Rbracket", value: "]", start: 5},
            ]);
    });
    it('should tokenize numbers with multiple digits', function() {
        assert.deepEqual(
            tokenize("12345"),
            [{type: "Number", value: 12345, start: 0}]);
    });
    it('should tokenize negative numbers', function() {
        assert.deepEqual(
            tokenize("-12345"),
            [{type: "Number", value: -12345, start: 0}]);
    });
    it('should tokenize quoted identifier', function() {
        assert.deepEqual(tokenize('"foo"'),
                         [{type: "QuotedIdentifier",
                          value: "foo",
                          start: 0}]);
    });
    it('should tokenize quoted identifier with unicode escape', function() {
        assert.deepEqual(tokenize('"\\u2713"'),
                         [{type: "QuotedIdentifier",
                          value: "✓",
                          start: 0}]);
    });
    it('should tokenize literal lists', function() {
        assert.deepEqual(tokenize("`[0, 1]`"),
                         [{type: "Literal",
                          value: [0, 1],
                          start: 0}]);
    });
    it('should tokenize literal dict', function() {
        assert.deepEqual(tokenize("`{\"foo\": \"bar\"}`"),
                         [{type: "Literal",
                          value: {"foo": "bar"},
                          start: 0}]);
    });
    it('should tokenize literal strings', function() {
        assert.deepEqual(tokenize("`\"foo\"`"),
                         [{type: "Literal",
                          value: "foo",
                          start: 0}]);
    });
    it('should tokenize json literals', function() {
        assert.deepEqual(tokenize("`true`"),
                         [{type: "Literal",
                          value: true,
                          start: 0}]);
    });
    it('should not requiring surrounding quotes for strings', function() {
        assert.deepEqual(tokenize("`foo`"),
                         [{type: "Literal",
                          value: "foo",
                          start: 0}]);
    });
    it('should not requiring surrounding quotes for numbers', function() {
        assert.deepEqual(tokenize("`20`"),
                         [{type: "Literal",
                           value: 20,
                           start: 0}]);
    });
    it('should tokenize literal lists with chars afterwards', function() {
        assert.deepEqual(
            tokenize("`[0, 1]`[0]"), [
                {type: "Literal", value: [0, 1], start: 0},
                {type: "Lbracket", value: "[", start: 8},
                {type: "Number", value: 0, start: 9},
                {type: "Rbracket", value: "]", start: 10}
        ]);
    });
    it('should tokenize two char tokens with shared prefix', function() {
        assert.deepEqual(
            tokenize("[?foo]"),
            [{type: "Filter", value: "[?", start: 0},
             {type: "UnquotedIdentifier", value: "foo", start: 2},
             {type: "Rbracket", value: "]", start: 5}]
        );
    });
    it('should tokenize flatten operator', function() {
        assert.deepEqual(
            tokenize("[]"),
            [{type: "Flatten", value: "[]", start: 0}]);
    });
    it('should tokenize comparators', function() {
        assert.deepEqual(tokenize("<"),
                         [{type: "LT",
                          value: "<",
                          start: 0}]);
    });
    it('should tokenize two char tokens without shared prefix', function() {
        assert.deepEqual(
            tokenize("=="),
            [{type: "EQ", value: "==", start: 0}]
        );
    });
    it('should tokenize not equals', function() {
        assert.deepEqual(
            tokenize("!="),
            [{type: "NE", value: "!=", start: 0}]
        );
    });
    it('should tokenize the OR token', function() {
        assert.deepEqual(
            tokenize("a||b"),
            [
                {type: "UnquotedIdentifier", value: "a", start: 0},
                {type: "Or", value: "||", start: 1},
                {type: "UnquotedIdentifier", value: "b", start: 3}
            ]
        );
    });
    it('should tokenize function calls', function() {
        assert.deepEqual(
            tokenize("abs(@)"),
            [
                {type: "UnquotedIdentifier", value: "abs", start: 0},
                {type: "Lparen", value: "(", start: 3},
                {type: "Current", value: "@", start: 4},
                {type: "Rparen", value: ")", start: 5}
            ]
        );
    });

});


describe('parsing', function() {
    it('should parse field node', function() {
        assert.deepEqual(compile('foo'),
                          {type: 'Field', name: 'foo'});
    });
});

describe('strictDeepEqual', function() {
    it('should compare scalars', function() {
        assert.strictEqual(strictDeepEqual('a', 'a'), true);
    });
    it('should be false for different types', function() {
        assert.strictEqual(strictDeepEqual('a', 2), false);
    });
    it('should be false for arrays of different lengths', function() {
        assert.strictEqual(strictDeepEqual([0, 1], [1, 2, 3]), false);
    });
    it('should be true for identical arrays', function() {
        assert.strictEqual(strictDeepEqual([0, 1], [0, 1]), true);
    });
    it('should be true for nested arrays', function() {
        assert.strictEqual(
            strictDeepEqual([[0, 1], [2, 3]], [[0, 1], [2, 3]]), true);
    });
    it('should be true for nested arrays of strings', function() {
        assert.strictEqual(
            strictDeepEqual([["a", "b"], ["c", "d"]],
                            [["a", "b"], ["c", "d"]]), true);
    });
    it('should be false for different arrays of the same length', function() {
        assert.strictEqual(strictDeepEqual([0, 1], [1, 2]), false);
    });
    it('should handle object literals', function() {
        assert.strictEqual(strictDeepEqual({a: 1, b: 2}, {a: 1, b: 2}), true);
    });
    it('should handle keys in first not in second', function() {
        assert.strictEqual(strictDeepEqual({a: 1, b: 2}, {a: 1}), false);
    });
    it('should handle keys in second not in first', function() {
        assert.strictEqual(strictDeepEqual({a: 1}, {a: 1, b: 2}), false);
    });
    it('should handle nested objects', function() {
        assert.strictEqual(
            strictDeepEqual({a: {b: [1, 2]}},
                            {a: {b: [1, 2]}}), true);
    });
    it('should handle nested objects that are not equal', function() {
        assert.strictEqual(
            strictDeepEqual({a: {b: [1, 2]}},
                            {a: {b: [1, 4]}}), false);
    });
});

describe('search', function() {
    it('should by default support case sensitive comparison via compatators', function() {
        assert.equal(jmespath.search({foo: 'bar'}, "foo == 'BAR'"), false);
        assert.equal(jmespath.search({foo: 'BAR'}, "foo == 'bar'"), false);
        assert.equal(jmespath.search({foo: 'bar'}, 'contains(`["BAR"]`, foo)'), false);
        assert.equal(jmespath.search({foo: 'bar'}, '`{"bar": "baz", "qux": 2}`.bar == \'BAZ\''), false);
        assert.equal(jmespath.search({foo: 'bar'}, "foo != 'BAR'"), true);
        assert.equal(jmespath.search({foo: 'BAR'}, "foo != 'bar'"), true);
        assert.equal(jmespath.search({foo: 2}, 'contains(`[1, 2, 3]`, foo)'), true);
        assert.equal(jmespath.search({foo: 2}, '`{"bar": 2}`.bar == foo'), true);
    });
    it('should support case insensitive comparison via compatators', function() {
        var opts = { useCaseInsensitiveComparison: true };
        assert.equal(jmespath.search({foo: 'bar'}, "foo == 'BAR'", opts), true);
        assert.equal(jmespath.search({foo: 'BAR'}, "foo == 'bar'", opts), true);
        assert.equal(jmespath.search({foo: 'bar'}, 'contains(`["BAR"]`, foo)', opts), true);
        assert.equal(jmespath.search({foo: 'bar'}, '`{"bar": "baz", "qux": 2}`.bar == \'BAZ\'', opts), true);
        assert.equal(jmespath.search({foo: 'bar'}, "foo != 'BAR'", opts), false);
        assert.equal(jmespath.search({foo: 'BAR'}, "foo != 'bar'", opts), false);
        assert.equal(jmespath.search({foo: 2}, 'contains(`[1, 2, 3]`, foo)', opts), true);
        assert.equal(jmespath.search({foo: 2}, '`{"bar": 2}`.bar == foo', opts), true);
    });
    it('should by default support case sensitive comparison via functions', function() {
        assert.equal(jmespath.search({foo: 'bar'}, "starts_with(foo, 'B')"), false);
        assert.equal(jmespath.search({foo: 'bar'}, "ends_with(foo, 'AR')"), false);
    });
    it('should support case insensitive comparison via functions', function() {
        var opts = { useCaseInsensitiveComparison: true };
        assert.equal(jmespath.search({foo: 'bar'}, "starts_with(foo, 'B')", opts), true);
        assert.equal(jmespath.search({foo: 'bar'}, "ends_with(foo, 'AR')", opts), true);
    });
    it(
        'should throw a readable error when invalid arguments are provided to a function',
        function() {
            try {
                jmespath.search([], 'length(`null`)');
            } catch (e) {
                assert(e.message.search(
                    'expected argument 1 to be type string,array,object'
                ), e.message);
                assert(e.message.search('received type null'), e.message);
            }
        }
    );
});

describe('over', function() {
  it('should mutate objects', function() {
    assert.deepStrictEqual(
      jmespath.over({a: 1, b: 2}, 'a', (a) => a.toString()),
      {a: '1', b: 2}
    );
    assert.deepStrictEqual(
      jmespath.over({a: {b: {c: {d: 1}}}, e: 2}, 'a.b.c.d', (a) => a.toString()),
      {a: {b: {c: {d: '1'}}}, e: 2}
    );
  });

  it('should mutate arrays', function() {
    assert.deepStrictEqual(
      jmespath.over({a: [1, 2, 3]}, 'a | [0]', (a) => a.toString()),
      {a: ['1', 2, 3]}
    );
  });

  it('should mutate slices', function() {
    assert.deepStrictEqual(
      jmespath.over({a: [1, 2, 3]}, 'a[0:3]', (a) => a.toString()),
      {a: ['1', '2', '3']}
    );
  });
});

describe('remove', function() {
    it('should remove a basic traversal', function() {
        let testData = {
            a: {
                "b": {
                    "c": [1, 2, 3]
                }
            }
        }
        assert.deepStrictEqual(
            jmespath.remove(testData, 'a.b.c', (a) => a.toString()),
            {a: {b: {}}}
        );
    });

    it('should remove a nested list traversal', function() {
        let testData = {
            a: [
                {
                    "b": {
                        "c": [1, 2, 3]
                    }
                }
            ]
        }
        assert.deepStrictEqual(
            jmespath.remove(testData, 'a[0].b.c', (a) => a.toString()),
            {a: [{b: {}}]}
        );
    });
    it('should NOT remove a nested list', function() {
        let testData = {
            a: [
                {
                    "b": {
                        "c": [[1,2,3], 2, 3]
                    }
                }
            ]
        }
        assert.deepStrictEqual(
            jmespath.remove(testData, 'a[0].b.c[0]', (a) => a.toString()),
            {a: [{b: {"c": [[1,2,3], 2, 3]}}]}
        );
    });
    it('should NOT remove a flatten', function() {
        let testData = {
            a: [
                [{"a": "b"}],
                "c"
            ]
        }
        assert.deepStrictEqual(
            jmespath.remove(testData, 'a[]', (a) => a.toString()),
            {a:[[{"a":"b"}], "c"]}
        );
    });
    it('should NOT remove a flatten index', function() {
        let testData = {
            a: [
                [{"a": "b"}],
                "c"
            ]
        }
        assert.deepStrictEqual(
            jmespath.remove(testData, 'a[][0]', (a) => a.toString()),
            {a:[[{"a":"b"}], "c"]}
        );
    });
    it('should NOT remove a flatten slice', function() {
        let testData = {
            a: [
                [{"a": "b"}],
                "c"
            ]
        }
        assert.deepStrictEqual(
            jmespath.remove(testData, 'a[][0:]', (a) => a.toString()),
            {a:[[{"a":"b"}], "c"]}
        );
    });
});

describe('decorate', function() {
  it(
    'should call a custom function when called via decorator',
    function() {
      var TYPE_NUMBER = 0;
      function customFunc(resolvedArgs) {
        return resolvedArgs[0] + 99;
      }
      var extraFunctions = {
        custom: {_func: customFunc, _signature: [{types: [TYPE_NUMBER]}]},
      };
      var value = jmespath.decorate(extraFunctions)('custom(`1`)')({});
      assert.strictEqual(value, 100);
    }
  );
  it(
    'should provide a compiled expression that can be cached and reused',
    function() {
      var expr = jmespath.decorate({})('a');
      var value = expr({ a: 1 });
      assert.strictEqual(value, 1);
      value = expr({ a: 2 });
      assert.strictEqual(value, 2);
    }
  );
});

describe('root', function() {
  it(
    '$ should give access to the root value',
    function() {
      var value = jmespath.search({ foo: { bar: 1 }}, 'foo.{ value: $.foo.bar }');
      assert.equal(value.value, 1);
    }
  );
  it(
    '$ should give access to the root value after pipe',
    function() {
      var value = jmespath.search({ foo: { bar: 1 }}, 'foo | $.foo.bar');
      assert.strictEqual(value, 1);
    }
  );
});


test/compliance.js
var fs = require('fs');
var path = require('path');
var assert = require('assert');
var jmespath = require('../src/jmespath');
var search = jmespath.search;

// Compliance tests that aren't supported yet.
var notImplementedYet = [];

function endsWith(str, suffix) {
    return str.indexOf(suffix, str.length - suffix.length) !== -1;
}

var listing = fs.readdirSync('test/compliance');
for (var i = 0; i < listing.length; i++) {
    var filename = 'test/compliance/' + listing[i];
    if (fs.statSync(filename).isFile() && endsWith(filename, '.json') &&
        notImplementedYet.indexOf(path.basename(filename)) === -1) {
        addTestSuitesFromFile(filename);
    }
}
function addTestSuitesFromFile(filename) {
    describe(filename, function() {
        var spec = JSON.parse(fs.readFileSync(filename, 'utf-8'));
        var errorMsg;
        for (var i = 0; i < spec.length; i++) {
            var msg = "suite " + i + " for filename " + filename;
            describe(msg, function() {
                var given = spec[i].given;
                var cases = spec[i].cases;
                for (var j = 0; j < cases.length; j++) {
                    var testcase = cases[j];
                    if (testcase.error !== undefined) {
                        // For now just verify that an error is thrown
                        // for error tests.
                        (function(testcase, given) {
                          it('should throw error for test ' + j, function() {
                              assert.throws(
                                  function() {
                                    search(given, testcase.expression);
                                  }, Error, testcase.expression);
                          });
                        })(testcase, given);
                    } else {
                        (function(testcase, given) {
                          it('should pass test ' + j + " expression: " + testcase.expression, function() {
                              assert.deepEqual(search(given, testcase.expression),
                                               testcase.result);
                          });
                        })(testcase, given);
                    }
                }
            });
        }
    });
}


perf.js
var jmespath = require('./src/jmespath')
var Benchmark = require('benchmark');
var suite = new Benchmark.Suite;

// add tests
suite.add('Parser#single_expr', function() {
  jmespath.compile("foo");
})
.add('Parser#single_subexpr', function() {
  jmespath.compile("foo.bar");
})
.add('Parser#deeply_nested_50', function() {
  jmespath.compile("j49.j48.j47.j46.j45.j44.j43.j42.j41.j40.j39.j38.j37.j36.j35.j34.j33.j32.j31.j30.j29.j28.j27.j26.j25.j24.j23.j22.j21.j20.j19.j18.j17.j16.j15.j14.j13.j12.j11.j10.j9.j8.j7.j6.j5.j4.j3.j2.j1.j0");

})
.add('Parser#deeply_nested_50_index', function() {
  jmespath.compile("[49][48][47][46][45][44][43][42][41][40][39][38][37][36][35][34][33][32][31][30][29][28][27][26][25][24][23][22][21][20][19][18][17][16][15][14][13][12][11][10][9][8][7][6][5][4][3][2][1][0]");
})
.add('Parser#basic_list_projection', function() {
  jmespath.compile("foo[*].bar");
})
.on('cycle', function(event) {
  var bench = event.target;
  var mean = bench.stats.mean * 1000;
  var variance = bench.stats.variance * 1000000;
  var result = 'Mean time: ' + mean.toFixed(6) + 'msec ';
  result += event.target.toString();
  console.log(result);
})
.on('complete', function() {
})
// run async
.run({ 'async': false });
